scipy/signal/_spectral_py.py

```diff
@@ -586,29 +589,39 @@ def csd(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
     >>> plt.show()
 
     """
+    xp = array_namespace(x)
     freqs, _, Pxy = _spectral_helper(x, y, fs, window, nperseg, noverlap,
                                      nfft, detrend, return_onesided, scaling,
                                      axis, mode='psd')
 
     # Average over windows.
-    if len(Pxy.shape) >= 2 and Pxy.size > 0:
+    if len(Pxy.shape) >= 2 and size(Pxy) > 0:
         if Pxy.shape[-1] > 1:
             if average == 'median':
-                # np.median must be passed real arrays for the desired result
+                # xp.median must be passed real arrays for the desired result
                 bias = _median_bias(Pxy.shape[-1])
-                if np.iscomplexobj(Pxy):
-                    Pxy = (np.median(np.real(Pxy), axis=-1)
-                           + 1j * np.median(np.imag(Pxy), axis=-1))
+                if Pxy.dtype in [xp.complex64, xp.complex128]:
+                    Pxy = (xp.median(xp.real(Pxy), axis=-1)
+                           + 1j * xp.median(xp.imag(Pxy), axis=-1))
                 else:
-                    Pxy = np.median(Pxy, axis=-1)
-                Pxy /= bias
+                    Pxy = xp.median(Pxy, axis=-1)
+                # for PyTorch, Pxy is torch.return_types.median
```

Comment:
Looks like we need to write an array-agnostic `median` à la `cov` (or push for it to be added to the standard). Again, I would remove the special casing for now and let this fail so that it's clear where work needs to be done.